pkg/controller/elasticsearch/client/v6.go

```diff
@@ -59,9 +59,15 @@ func (c *clientV6) Flush(ctx context.Context) error {
 	return c.post(ctx, "/_flush", nil, nil)
 }
 
-func (c *clientV6) GetClusterHealth(ctx context.Context) (Health, error) {
+func (c *clientV6) GetClusterHealth(ctx context.Context, params url.Values) (Health, error) {
 	var result Health
-	return result, c.get(ctx, "/_cluster/health", &result)
+	err := c.get(ctx, "/_cluster/health?"+params.Encode(), &result)
+	if IsTimeout(err) {
+		// ignore timeout errors as they are communicated in the returned payload so we can reserve error handling
```

Comment:
it's in the payload, but do we ever want to treat a timeout as *not* an error condition? I see some of the other callers we have don't check for timeouts but probably should? example: https://github.com/elastic/cloud-on-k8s/pull/3195/files#diff-972e1e28f90c08fd73595db4b03025d2L73

Reply:
I believe those would happen on an idle connection in the pool and are handled by the Go http transport thingy IIUC https://go-review.googlesource.com/c/go/+/179457/4/src/net/http/transport.go#1931

But I like the idea of a separate API, my only concern is maybe that we are mixing levels of abstraction. So far all client APIs have been direct representations of Elasticsearch REST APIs this would be the first 'higher-level' API. My gut feeling is that we should do this in layers: have a low-level client that just exposes ES APIs and then a higher-level client interface on top that offers these more convenient abstractions. But maybe I am over-complicating things. wdyt?

Reply:
You're right about the abstraction level. `ClusterBootstrappedForZen2()` is another example of a function doing a bit more than API calls.

Splitting lower vs. higher level functions attached to different structs may make things a bit cleaner but I'm not 100% sure it would reduce complexity.

Getting the cluster health with additional params is still a low-level API call?
I think I'd be fine with something like these 2 functions if it simplifies our life:

```go
func (c *clientV6) GetClusterHealth(ctx context.Context) (Health, error) {
	var result Health
	return result, c.get(ctx, "/_cluster/health", &result)
}

func (c *clientV6) GetClusterHealthWaitForLanguid(ctx context.Context, timeout time.Time) (Health, error) {
	var result Health
	// wait for all "languid" events (lowest priority) to be processed, or time out
	url := fmt.Sprintf("/_cluster/health?wait_for_events=languid&timeout=%ds", timeout.Second())
	err := c.get(ctx, url, &result)
	if IsTimeout(err) && result.TimedOut{
		// timeout reached, which is not an unexpected error
		// we still have a response body to work with
		err = nil
	}
	return result, err
}
```